File: eval/generate.py
#!/usr/bin/env python3

import sys
import json
import torch

sys.path.append("..")
from arch import architecture

vocab_size = 128
filename = sys.argv[1]
max_tok = int(sys.argv[2])

model = architecture.GPTLanguageModel(vocab_size)
model.load_state_dict(torch.load(filename))
model.eval()

prompt = sys.stdin.read()
context = torch.tensor([[int(n) for n in prompt.split()]], dtype=torch.long)
if len(context[0]) == 0:
     context = torch.zeros((1, 1), dtype=torch.long)

tokens = model.generate(context, max_new_tokens=max_tok)[0].tolist()
print(" ".join(str(t) for t in tokens))

File: eval/generate.js
#!/usr/bin/env node

import * as tf from '@tensorflow/tfjs-node';
import { createInterface } from 'readline';
import { GPTLanguageModel } from '../arch/architecture.js';

// initialize arguments
const filePath = process.argv[2];
const maxTokens = parseInt(process.argv[3]);
const vocabSize = 128;

// initialize model
const model = new GPTLanguageModel(vocabSize);
model.build();
await model.load(filePath);

// handle stdin
async function main() {
    const rl = createInterface({input: process.stdin});
    let promptVal = '';
    rl.on('line', (line) => {
        promptVal += line;
    });
    rl.on('close', () => {
        runGeneration(promptVal);
    });
}

async function runGeneration(promptVal) {
    let context;

    if (promptVal.trim()) {
        const tokens = promptVal.trim().split(/\s+/).map(Number);
        context = tf.tensor2d([tokens], undefined, 'int32');
    } else {
        context = tf.zeros([1, 1], 'int32');
    }

    const generated = model.generate(context, maxTokens).arraySync()[0];
    console.log(generated.join(' '));

    context.dispose();
    //generated.dispose();
}

main();
